tofhir-server/src/main/scala/io/tofhir/server/service/ProjectService.scala

```diff
@@ -65,7 +75,65 @@ class ProjectService(projectRepository: IProjectRepository) extends LazyLogging
    * @return
    */
   def removeProject(id: String): Future[Unit] = {
-    projectRepository.removeProject(id)
+    val jobIds: Seq[String] = projectRepository.getJobIds(id)
+    val mappingIds: Seq[String] = projectRepository.getMappingIds(id)
+    val mappingContextIds: Seq[String] = projectRepository.getMappingContextIds(id)
+    val schemaIds: Seq[String] = projectRepository.getSchemaIds(id)
+
+    // Define future for job deletion
+    def jobDeletion = Future.traverse(jobIds)(jobId => {
+      // try catch is necessary because recover cannot detect thrown exceptions before future is returned
+      try{
+        jobRepository.deleteJob(id, jobId)
+      }catch {
+        case e: Throwable => Future.failed(e)
+      }
+    }.recover({
+      case e: Throwable => logger.error(s"Error while deleting the job: ${jobId}", e)
+    }))
+    // Define future for mapping deletion
+    def mappingDeletion = Future.traverse(mappingIds)(mappingId =>
+      {
+        try{
+          mappingRepository.deleteMapping(id, mappingId)
+        }catch {
+          case e: Throwable => Future.failed(e)
+        }
+      }.recover({
+      case e: Throwable => logger.error(s"Error while deleting the mapping: ${mappingId}", e)
+    }))
+    // define future for mapping context deletion
+    def mappingContextDeletion = Future.traverse(mappingContextIds)(mappingContextId =>
+      {
+        try {
+          mappingContextRepository.deleteMappingContext(id, mappingContextId)
+        }catch {
+          case e: Throwable => Future.failed(e)
+        }
+      }.recover({
+      case e: Throwable => logger.error(s"Error while deleting the mapping context: ${mappingContextId}", e)
+    }))
+    // define future for schema deletion
+    def schemaDeletion = Future.traverse(schemaIds)(schemaId =>
+      {
+        try {
+          schemaRepository.deleteSchema(id, schemaId)
+        }catch {
+          case e: Throwable => Future.failed(e)
+        }
+      }.recover({
+      case e: Throwable => logger.error(s"Error while deleting the schema: ${schemaId}", e)
+    }))
+    // Execute deletion futures in order
+    for {
```

Comment:
This creates an extra complexity in the ultithreaded executions. As far as I know, the for() yield structure creates all possible combinations of the enumerations that you give within the for block.

Instead I suggest to gow with somtthing like the following:

```
Future.sequence(jobIds.map(jobid => {
  jobRepository.deleteJob(jobid)
})).flatMap(f => {Future.sequence(mappingIds.map(mappingId => {...}))})
```